DigitalLearningSolutions.Data/Services/RecommendedLearningService.cs

```diff
@@ -45,27 +49,41 @@ ILearningLogItemsDataService learningLogItemsDataService
                 int delegateId
             )
         {
-            var competencyIds = selfAssessmentDataService.GetCompetencyIdsForSelfAssessment(selfAssessmentId);
+            var hasMaxSignpostedResources = Int32.TryParse(
+                configDataService.GetConfigValue(ConfigDataService.MaxSignpostedResources),
+                out var maxSignpostedResources
+            );
 
+            var competencyIds = selfAssessmentDataService.GetCompetencyIdsForSelfAssessment(selfAssessmentId);
             var competencyLearningResources = new List<CompetencyLearningResource>();
+
             foreach (var competencyId in competencyIds)
             {
                 var learningHubResourceReferencesForCompetency =
                     competencyLearningResourcesDataService.GetActiveCompetencyLearningResourcesByCompetencyId(
                         competencyId
                     );
+
                 competencyLearningResources.AddRange(learningHubResourceReferencesForCompetency);
             }
 
             var resourceReferences = competencyLearningResources.Select(
                 clr => (clr.LearningHubResourceReferenceId, clr.LearningResourceReferenceId)
-            ).Distinct().ToDictionary(x => x.LearningHubResourceReferenceId, x => x.LearningResourceReferenceId);
+            ).Distinct().ToDictionary(
+                x => x.LearningHubResourceReferenceId,
+                x => x.LearningResourceReferenceId
+            );
 
-            var uniqueLearningHubReferenceIds = competencyLearningResources
-                .Select(clr => clr.LearningHubResourceReferenceId).Distinct().ToList();
+            var uniqueLearningHubReferenceIds = competencyLearningResources.Select(
+                clr => clr.LearningHubResourceReferenceId
+            ).Distinct();
 
             var resources =
-                await learningHubResourceService.GetBulkResourcesByReferenceIds(uniqueLearningHubReferenceIds);
+                await learningHubResourceService.GetBulkResourcesByReferenceIds(
+                    hasMaxSignpostedResources
+                        ? uniqueLearningHubReferenceIds.Take(maxSignpostedResources).ToList()
+                        : uniqueLearningHubReferenceIds.ToList()
+                );

```

Comment:
Isn't this going to take 150 resources and then order them by score, instead of having the best 150 resources?

The problem was that the response from the method `GetBulkResourcesByReferenceIds` was too big when more than 150 resources were queried. I think we could make multiple queries for 150 resources each and then sort them all.

Reply:
No, the problem was that the _request to_ the learning hub API (as made in `GetBulkResourcesByReferenceIds`) was too big: the URL exceeded 2048 characters because of the format (`resourceReferenceIds=1&resourceReferenceIds=2&...`)

https://softwiretech.atlassian.net/browse/HEEDLS-1033 will resolve that, by changing the URL we use/having a different query string format, but the requirement of HEEDLS-1036 is still to limit to only 150 ids in the request. 

When imposing this limit, it is not possible request the "best" 150 resources - we only know recommendation score after we make the request. I agree that we _could_ make multiple queries to get all the data and then sort them all, but it seems unlikely that, in the situation where there are more than 150, the user will actually want to wade through even the first 150, let alone anything beyond that. (I seem to recall having a conversation with Steve where he was saying this, but can't find it in slack)

Reply:
I agree that a user will only go through the first few resources, but that's exactly why this method sorts by score, so those first few resources are actually the best. If it's not too much effort, I think we should have multiple requests, sort all resources and optionally just return the top few instead of all of them.